packages/backend-common/src/reading/tree/TarArchiveResponse.ts

```diff
@@ -44,6 +44,7 @@ export class TarArchiveResponse implements ReadTreeResponse {
     private readonly workDir: string,
     public readonly etag: string,
     private readonly filter?: (path: string, info: { size: number }) => boolean,
+    private readonly stripFirstDirectory: boolean = true,
```

Comment:
Is this default needed if the factory takes care of it? I would rather remove it here to handle it at one place only.

Reply:
on the other hand, it gets a breaking change for the constructor of this type then

Reply:
I'll also add that this is problematic. It's a container-specific property and "leaky". In fact, this type has had a strip parameter in the past and we worked to get rid of it - would like to see that it didn't need to get added back

Reply:
Yes you're right but I wanted to limit my changes to avoid breaking other providers :sweat_smile: 

It seems that the `ZipArchiveResponse` had the `stripFirstDirectory` in the past and now the behaviors of the two are different:
- `ZipArchiveResponse` doesn't strip the top level directory (but it's stripped by the `AzureUrlReader` using `subpath`)
- `TarArchiveResponse` strips the top level directory

I'll try to find a way so that both doesn't strip the top level directory, but depending on the provider we need to determine the top level directory name to construct the `subpath` accordingly.

Reply:
Some possible background on the Azure side of this: https://github.com/backstage/backstage/issues/4359

Reply:
I'm having trouble finding a way to remove the stripFirstDirectory, given that each provider that uses the `TarArchiveResponse` (GitLab, BitBucket, GitHub and now Gerrit) uses a different pattern for the first parent directory.

Also, it seems that test archives in `__fixtures__/` are not using the actual naming pattern.
For example, in GitLab.com for a repository `myorg/myrepo`, the archive contains a folder named `myrepo-main`.
The archive that is used for tests (`__fixtures__/gitlab-archive.tar.gz`) contains a folder `gitlab-archive`.

I could make the archive follow the actual pattern but I don't have the confidence that the pattern is correct and that this change won't actually introduce a bug